### PR TITLE
Set semantic background color for submenu items

### LINKED ISSUE #13: Color coding for radial items

### PR DESCRIPTION
**Pull request by jon-kroko**
closes #13

### CODE REVIEW

File: src/ColorContextMenu-Core.package/ColorChooserMorph.class/instance/buildDiscreteMenu.attribute.levels..st [hunk 1]
```diff
@@ -1,12 +1,18 @@
 menus
 buildDiscreteMenu: menuFactory attribute: aSymbol levels: levels
 
-	| menu setter |
+	| menu setter getter steps item |
+	self flag: 'Warum heißt es hier menuFactory und nicht aRadialMenuMorph?'.
```
Comment (LinqLover) [inline src/ColorContextMenu-Core.package/ColorChooserMorph.class/instance/buildDiscreteMenu.attribute.levels..st]: please remove comment again

File: src/ColorContextMenu-Core.package/ColorChooserMorph.class/instance/buildDiscreteMenu.attribute.levels..st [hunk 2]
```diff
@@ -1,12 +1,18 @@
...
-		menu add: ('{1}%' format: {(level / (levels - 1) * 100) rounded}) action: [
-			self perform: setter with: level / (levels - 1)]].
+	steps := levels - 1.
+	0 to: steps do: [:level |
+		item := menu 
+			add: ('{1}%' format: {(level / steps * 100) rounded}) 
+			action: [self perform: setter with: level / steps].
+		item class = RadialMenuItemMorph ifTrue: [
+			getter := ('get{1}:' format: {aSymbol capitalized}) asSymbol.
+			item baseColor: (self perform: getter with: level / steps)]].
```
Comment (LinqLover) [inline src/ColorContextMenu-Core.package/ColorChooserMorph.class/instance/buildDiscreteMenu.attribute.levels..st]: Let's rework this to respect the `#colorizeMenuItem:color:` change from https://github.com/LinqLover/ColorContextMenu/pull/17/commits/5cd15b23516e35a6306c6141afb2997dd4ef4bb3 when it's ready :)

Apart from that, instead of `item class = RadialMenuItemMorph` always write `(item isKindOf: RadialMenuItemMorph)` :) We should refactor this anyway later ...

File: src/ColorContextMenu-Core.package/ColorChooserMorph.class/instance/defaultNumberOfLevels.st [hunk 3]
```diff
@@ -1,4 +1,4 @@
-private
+as yet unclassified
```
Comment (LinqLover) [inline src/ColorContextMenu-Core.package/ColorChooserMorph.class/instance/defaultNumberOfLevels.st]: Please revert

```suggestion
private
```

File: src/ColorContextMenu-Core.package/ColorChooserMorph.class/instance/referenceColorForItems.st [hunk 4]
```diff
@@ -0,0 +1,4 @@
+as yet unclassified
```
Comment (LinqLover) [inline src/ColorContextMenu-Core.package/ColorChooserMorph.class/instance/referenceColorForItems.st]: Recategorize

```suggestion
private
```

File: src/ColorContextMenu-Core.package/RadialMenuItemMorph.class/instance/updateLabelColor.st [hunk 5]
```diff
@@ -0,0 +1,6 @@
+as yet unclassified
```
Comment (LinqLover) [inline src/ColorContextMenu-Core.package/RadialMenuItemMorph.class/instance/updateLabelColor.st]:
```suggestion
updating
```

File: src/ColorContextMenu-Core.package/RadialMenuItemMorph.class/instance/updateLabelColor.st [hunk 6]
```diff
@@ -0,0 +1,6 @@
...
+updateLabelColor
+
+	self color luminance > 0.5 
+		ifTrue: [self labelColor: Color black]
+		ifFalse: [self labelColor: Color white]
```
Comment (LinqLover) [inline src/ColorContextMenu-Core.package/RadialMenuItemMorph.class/instance/updateLabelColor.st]: Funny, this almost exactly what `#makeForegroundColor` does 😅

```suggestion
	self labelColor: self color makeForegroundColor.
```

File: src/ColorContextMenu-Core.package/RadialMenuMorph.class/instance/outerFractionForItems.st [hunk 7]
```diff
@@ -1,4 +1,4 @@
 accessing
 outerFractionForItems
 
-	^ self outerFraction - 0.05
\ No newline at end of file
+	^ self outerFraction "- 0.05"
```
Comment (LinqLover) [inline src/ColorContextMenu-Core.package/RadialMenuMorph.class/instance/outerFractionForItems.st]: Hm ... I used to like the border. Why did you remove it?

File: src/ColorContextMenu-Core.package/RadialMenuItemMorph.class/instance/update..st [hunk 8]
```diff
@@ -0,0 +1,4 @@
+as yet unclassified
+update: aSymbol
+
+	aSymbol = #color ifTrue: [self updateLabelColor]
```
Comment (LinqLover) [inline src/ColorContextMenu-Core.package/RadialMenuItemMorph.class/instance/update..st]: In most cases, it is an anti-pattern to use observer on yourself. Can't you simply override the `#color:` setter in this case? Also affects the changes in `#privateInteractionState:` and `#initialize`.

File: src/ColorContextMenu-Core.package/RadialMenuItemMorph.class/instance/initialize.st [hunk 9]
```diff
@@ -5,6 +5,7 @@ initialize
 	
 	self privateInteractionState: nil.
 	self borderWidth: self defaultBorderWidth.
+	self addDependent: self.
 	
 	self addMorph: (labelMorph := self basicLabelMorph).
-	labelMorph color: Color white.
\ No newline at end of file
+	"labelMorph color: Color white."
```
Comment (LinqLover) [inline src/ColorContextMenu-Core.package/RadialMenuItemMorph.class/instance/initialize.st]: Is this WIP or can you remove this line?

File: src/ColorContextMenu-Core.package/ColorChooserMorph.class/instance/setHue..st [hunk 10]
```diff
@@ -1,4 +1,4 @@
 actions
 setHue: aFraction
 
-	self color: (self color ccmWithHue: aFraction).
\ No newline at end of file
+	self color: (self color ccmWithHue: aFraction)
```
Comment (LinqLover) [inline src/ColorContextMenu-Core.package/ColorChooserMorph.class/instance/setHue..st]: Alright, we seem to have different coding styles, but switching back and forth again and again won't solve this for sure 😅

File: src/ColorContextMenu-Core.package/ColorChooserMorph.class/instance/getBlack..st [hunk 11]
```diff
@@ -0,0 +1,4 @@
+accessing
+getBlack: aFraction
+
+	^ Color white ccmWithCmykBlack: aFraction
```
Comment (LinqLover) [inline src/ColorContextMenu-Core.package/ColorChooserMorph.class/instance/getBlack..st]: To all the new get selectors: What about extracting these magic numbers (white and black) to something like `#referenceColorRgbForItems` and `#referenceColorCmyk`? Took me a while to find out where these values come from ^^

**Review by LinqLover — CHANGES_REQUESTED** [review]
Looks good, some minor comments only :)

File: src/ColorContextMenu-Core.package/ColorChooserMorph.class/instance/getBlack..st (hunk 11 above)
Comment (jon-kroko) [reply src/ColorContextMenu-Core.package/ColorChooserMorph.class/instance/getBlack..st]: done

File: src/ColorContextMenu-Core.package/ColorChooserMorph.class/instance/setHue..st (hunk 10 above)
Comment (jon-kroko) [reply src/ColorContextMenu-Core.package/ColorChooserMorph.class/instance/setHue..st]: was on mistake, reverted

File: src/ColorContextMenu-Core.package/RadialMenuItemMorph.class/instance/initialize.st (hunk 9 above)
Comment (jon-kroko) [reply src/ColorContextMenu-Core.package/RadialMenuItemMorph.class/instance/initialize.st]: removed

File: src/ColorContextMenu-Core.package/RadialMenuItemMorph.class/instance/update..st (hunk 8 above)
Comment (jon-kroko) [reply src/ColorContextMenu-Core.package/RadialMenuItemMorph.class/instance/update..st]: done

File: src/ColorContextMenu-Core.package/RadialMenuMorph.class/instance/outerFractionForItems.st (hunk 7 above)
Comment (jon-kroko) [reply src/ColorContextMenu-Core.package/RadialMenuMorph.class/instance/outerFractionForItems.st]: mostly because the border between top-level menu and submenu was twice as thick as the outer borders (see image below)... also i just like the material look :)

File: src/ColorContextMenu-Core.package/RadialMenuItemMorph.class/instance/updateLabelColor.st (hunk 5 above)
Comment (jon-kroko) [reply src/ColorContextMenu-Core.package/RadialMenuItemMorph.class/instance/updateLabelColor.st]: done

File: src/ColorContextMenu-Core.package/ColorChooserMorph.class/instance/referenceColorForItems.st (hunk 4 above)
Comment (jon-kroko) [reply src/ColorContextMenu-Core.package/ColorChooserMorph.class/instance/referenceColorForItems.st]: done

File: src/ColorContextMenu-Core.package/ColorChooserMorph.class/instance/defaultNumberOfLevels.st (hunk 3 above)
Comment (jon-kroko) [reply src/ColorContextMenu-Core.package/ColorChooserMorph.class/instance/defaultNumberOfLevels.st]: done

File: src/ColorContextMenu-Core.package/ColorChooserMorph.class/instance/buildDiscreteMenu.attribute.levels..st (hunk 2 above)
Comment (jon-kroko) [reply src/ColorContextMenu-Core.package/ColorChooserMorph.class/instance/buildDiscreteMenu.attribute.levels..st]: done the second part (isKindOf:)

File: src/ColorContextMenu-Core.package/ColorChooserMorph.class/instance/buildDiscreteMenu.attribute.levels..st (hunk 1 above)
Comment (jon-kroko) [reply src/ColorContextMenu-Core.package/ColorChooserMorph.class/instance/buildDiscreteMenu.attribute.levels..st]: done

File: src/ColorContextMenu-Core.package/RadialMenuMorph.class/instance/outerFractionForItems.st (hunk 7 above)
Comment (jon-kroko) [reply src/ColorContextMenu-Core.package/RadialMenuMorph.class/instance/outerFractionForItems.st]: Bildschirmfoto 2021-02-23 um 21 00 02

File: src/ColorContextMenu-Core.package/ColorChooserMorph.class/instance/buildDiscreteMenu.attribute.levels..st (hunk 2 above)
Comment (jon-kroko) [reply src/ColorContextMenu-Core.package/ColorChooserMorph.class/instance/buildDiscreteMenu.attribute.levels..st]: moved to new issue

File: .github/workflows/main.yml [hunk 12]
```diff
@@ -2,6 +2,7 @@ name: smalltalkCI
 
 on:
   push:
+  pull_request:
```
Comment (LinqLover) [inline .github/workflows/main.yml]: Merge fault!

File: src/ColorContextMenu-Core.package/Color.extension/class/ccmHue.saturation.brightness.alpha..st [hunk 13]
```diff
@@ -1,4 +1,4 @@
 *ColorContextMenu-Core-instance creation
 ccmHue: hue saturation: saturation brightness: brightness alpha: alpha
 
-	^ self h: hue * 360 s: saturation v: brightness alpha: alpha
\ No newline at end of file
+	^ self h: hue * 360 s: (saturation max: 0.005) v: (brightness max: 0.005) alpha: alpha
```
Comment (LinqLover) [inline src/ColorContextMenu-Core.package/Color.extension/class/ccmHue.saturation.brightness.alpha..st]: Is this a hack to preserve hue always? If yes, please move this change into `ColorChooserMorph` and add a small comment about it :)

File: src/ColorContextMenu-Core.package/ColorChooserMorph.class/instance/buildDiscreteMenu.attribute.levels..st [hunk 14]
```diff
@@ -2,7 +2,7 @@ menus
 buildDiscreteMenu: menuFactory attribute: aSymbol levels: levels
 
 	| menu setter getter steps item |
-	self flag: 'Warum heißt es hier menuFactory und nicht aRadialMenuMorph?'.
+
```
Comment (LinqLover) [inline src/ColorContextMenu-Core.package/ColorChooserMorph.class/instance/buildDiscreteMenu.attribute.levels..st]:
```suggestion
```

File: src/ColorContextMenu-Core.package/RadialMenuItemMorph.class/instance/color..st [hunk 15]
```diff
@@ -0,0 +1,5 @@
+as yet unclassified
```
Comment (LinqLover) [inline src/ColorContextMenu-Core.package/RadialMenuItemMorph.class/instance/color..st]: please recategorize

File: src/ColorContextMenu-Core.package/RadialMenuMorph.class/instance/outerFractionForItems.st (hunk 7 above)
Comment (jon-kroko) [reply src/ColorContextMenu-Core.package/RadialMenuMorph.class/instance/outerFractionForItems.st]: moved to #22

File: .github/workflows/main.yml (hunk 12 above)
Comment (LinqLover) [reply .github/workflows/main.yml]:
```suggestion
```

Comment (LinqLover) [inline .github/workflows/main.yml]:
```suggestion
```
